File: DotNetCorePracticalView/CS.DotNetCore.LoadTest/src/CS.DotNetCore.LoadTest.WebApp/Data/Mysql/DataMigrationMysqlDAO.cs
namespace CS.DotNetCore.LoadTest.WebApp.Data.Mysql
{
    using System;
    using Config;
    using MySql.Data.MySqlClient;
    using Schema;
    using Util.Data;
    using System.Text;

    internal class DataMigrationMysqlDAO : BaseDataMigrationDAO, IDataMigrationDAO
    {
        private IDatabaseConfig _config;
        private MySqlConnection _connection;

        public override string LatestVersion { get { return "1.0.0"; } }
                
        internal DataMigrationMysqlDAO(IDatabaseConfig config)
        {
            _config = config;
            _connection = new MySqlConnection(config.AdminString);
            OpenConnection(_connection.Open);
            CreateDatabase();
            CreateSchemaVersion();
        }

        private void CreateDatabase()
        {
            //create database if not exists
            var sql = string.Format("create database if not exists `{0}` character set='{1}' collate='{2}';",
                _config.DataBase, _config.Ctype, _config.Collate);

            DataHelper.ExecuteNonQuery(_connection, sql);
        }

        private void CreateSchemaVersion()
        {
            //create table SchemaVersion if not exists
            var sql = string.Format("use {0}; create table if not exists `{1}` (`{2}` char(10) not null);",
                _config.DataBase, SchemaVersionSchema.Table, SchemaVersionSchema.VersionColumn);

            DataHelper.ExecuteNonQuery(_connection, sql);
        }

        public string GetCurrentVersion()
        {
            //resolve current version
            var sql = string.Format("select `{0}` from `{1}`;", SchemaVersionSchema.VersionColumn, SchemaVersionSchema.Table);
            var currentVersion = DataHelper.ExecuteScalar(_connection, sql);
            return currentVersion == null ? null : currentVersion.ToString();
        }

        public void MigrateSchemaVersion(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate Schema Version", fromVersion, toVersion))
            {
                return;
            }

            var parameters = new MySqlParameter[1]
            {
                new MySqlParameter(SchemaVersionSchema.VersionColumn, MySqlDbType.VarChar)
                {
                    Direction = System.Data.ParameterDirection.Input,
                    IsNullable = false,
                    Value = toVersion
                }
            };

            var sql = string.Format("update `{0}` set `{1}` = @{1};", SchemaVersionSchema.Table, SchemaVersionSchema.VersionColumn);
            var affected = DataHelper.ExecuteNonQuery(_connection, sql, parameters);

            if (affected == 0)
            {
                sql = string.Format("insert into `{0}` (`{1}`) values (@{1});", SchemaVersionSchema.Table, SchemaVersionSchema.VersionColumn);
                DataHelper.ExecuteNonQuery(_connection, sql, parameters);
            }
        }

        public void MigrateIdentity(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate Identity", fromVersion, toVersion))
            {
                return;
            }

            var sql = string.Format("create table `{0}` (`{1}` varchar(100) not null, `{2}` json not null); ",
                IdentitySchema.Table, IdentitySchema.IdentityNameColumn, IdentitySchema.DataColumn);

            sql += string.Format("alter table `{0}` add constraint {1} primary key using btree (`{2}` asc);",
                IdentitySchema.Table, IdentitySchema.IdentityNameIndex, IdentitySchema.IdentityNameColumn);

            DataHelper.ExecuteNonQuery(_connection, sql);
        }

        public void MigrateDbmsUsers(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate DBMS Users", fromVersion, toVersion))
            {
                return;
            }

            var sql = string.Format("create user '{0}' identified by '{1}' password expire never; ",
                _config.User, _config.Password);

            sql += string.Format("grant create temporary tables, create view, delete, execute, insert, select, update on `{0}`.* to '{1}';",
                _config.DataBase, _config.User);

            DataHelper.ExecuteNonQuery(_connection, sql);
        }

        public void MigrateTestResult(string fromVersion, string toVersion)
        {
            var sql = new StringBuilder();
            sql.AppendFormat("create table `{0}`", TestResultSchema.Table);

            sql.AppendFormat(" (`{0}` varchar(10) not null,", TestResultSchema.TestIdColumn);
            sql.AppendFormat(" `{0}` varchar(50) null,", TestResultSchema.LanguageColumn);
            sql.AppendFormat(" `{0}` json not null,", TestResultSchema.DataColumn);

            sql.AppendFormat(" constraint {0} unique(`{1}`,`{2}`));",
                TestResultSchema.TestResultTestIndex, TestResultSchema.TestIdColumn, TestResultSchema.LanguageColumn);

            DataHelper.ExecuteNonQuery(_connection, sql.ToString());
        }

        public void Dispose()
        {
            _config = null;
            _connection.Close();

            _connection.Dispose();
            _connection = null;
        }
    }
}

File: DotNetCorePracticalView/CS.NodeJS.LoadTest.WebApp/webapp/data/mysql/dataMigrationMysqlDAO.js
let mysql = require('mysql');
let schema = require('../loadTestSchema.js');
let appSettings = require('../../config/appsettings.js');

module.exports = {

	createAsync: (mysqlConfig) => {

		let dao = {
			_dbConfig: mysqlConfig ? mysqlConfig : appSettings.databaseConfig.mysql,
			_client: null,
		};

		dao._closeConnectionAsync = () => {

			if (!dao._client) {
				return new Promise((res, rej) => { res(); });
			}

			return new Promise((res, rej) => {

				dao._client.end((err) => {

					if (err) {
						rej(err);
						return;
					}

					dao._client = null;
					res();
				});
			});
		};

		dao._openConnectionAsync = () => {

			let retryCount = 0;
			let retryTimeout = null;

			return new Promise((res, rej) => {

				let connectCallback = (err) => {

					if (err) {

						if (retryCount == 2) {
							rej(err);
							return;
						}

						console.warn('Data Migration connect fail. try: ' + (retryCount + 1) + ' starting retry..');
						retryCount++;

						retryTimeout = setTimeout(() => {

							dao._client.destroy();
							dao._client = mysql.createConnection(dao._dbConfig.rootConn);
							dao._client.connect(connectCallback);
						}, 10000);

						return;
					}

					if (retryCount > 0) {
						console.info('Data Migartion connect success.');
					}

					if (retryTimeout) {
						clearTimeout(retryTimeout);
					}

					res();
				}

				dao._client = mysql.createConnection(dao._dbConfig.rootConn);
				dao._client.connect(connectCallback);
			});
		};

		dao._createDatabaseAsync = () => {

			let sql = 'create database if not exists `' + dao._dbConfig.conn.database + '`';
			sql += ' character set=\'' + dao._dbConfig.ctype + '\'';
			sql += ' collate=\'' + dao._dbConfig.collate + '\';';

			let mysqlCmd = mysql.format(sql, []);

			return new Promise((res, rej) => {

				dao._openConnectionAsync()
					.then(() => {

						dao._client.query(mysqlCmd, (err, result) => {

							if (err) {
								rej(err)
								return;
							}

							res();
						});
					})
					.catch(rej);
			});
		};

		dao._createSchemaVersionAsync = () => {

			let sql = 'use ' + dao._dbConfig.conn.database + ';';
			sql += ' create table if not exists `' + schema.schemaVersion.table + '`';
			sql += ' (`' + schema.schemaVersion.versionColumn + '` char(10) not null);';

			let mysqlCmd = mysql.format(sql);

			return new Promise((res, rej) => {

				dao._openConnectionAsync()
					.then(() => {

						dao._client.query(mysqlCmd, (err, result) => {

							if (err) {
								rej(err);
								return;
							}

							res();
						});
					})
					.catch(rej);
			});
		};

		dao._validateMigrationOperation = (operation, fromVersion, toVersion) => {

			if (fromVersion || toVersion != dao.getLatestVersion()) {
				throw new Error(operation + ' from version "' + fromVersion + '"' + ' to version "' + toVersion + '" not implemented.');
			}

			if (fromVersion == toVersion) {
				console.log(operation + " - Updated");
				return false;
			}

			return true;
		}

		dao._updateSchemaVersionASync = (toVersion) => {

			let sql = 'update `' + schema.schemaVersion.table + '`';
			sql += ' set `' + schema.schemaVersion.versionColumn + '`=?;'

			let mysqlCmd = mysql.format(sql, [toVersion]);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, result) => {

					if (err) {
						rej(err);
						return;
					}

					res(result.affectedRows);
				})
			});
		};

		dao._insertSchemaVersionAsync = (toVersion) => {

			let sql = 'insert into `' + schema.schemaVersion.table + '`';
			sql += ' (`' + schema.schemaVersion.versionColumn + '`)';
			sql += ' values (?)';

			let mysqlCmd = mysql.format(sql, [toVersion]);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, result) => {

					if (err) {
						rej(err);
						return;
					}

					res();
				});
			});
		};

		dao.disposeAsync = () => {
			return dao._closeConnectionAsync();
		};

		dao.getCurrentVersionAsync = () => {

			let sql = 'select `' + schema.schemaVersion.versionColumn + '`';
			sql += ' from `' + schema.schemaVersion.table + '`;'

			let mysqlCmd = mysql.format(sql);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, rows) => {

					if (err) {
						rej(err);
						return;
					}

					if (rows && rows[0]) {
						res(rows[0][schema.schemaVersion.versionColumn]);
						return;
					}

					res(null);
				});
			});
		};

		dao.migrateSchemaVersionAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Schema Version', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			return new Promise((res, rej) => {

				dao._updateSchemaVersionASync(toVersion)
					.then((affected) => {

						if (!affected || affected < 1) {

							dao._insertSchemaVersionAsync(toVersion)
								.then(res)
								.catch(rej);

							return;
						}

						res();
					})
					.catch(rej);
			});
		}

		dao.migrateIdentityAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Identity', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			let sql = 'create table `' + schema.identity.table + '`';
			sql += ' (`' + schema.identity.identityNameColumn + '` varchar(100) not null,';
			sql += ' `' + schema.identity.dataColumn + '` json not null);';

			sql += ' alter table `' + schema.identity.table + '`';
			sql += ' add constraint ' + schema.identity.IdentityNameIndex + ' primary key';
			sql += ' using btree (`' + schema.identity.identityNameColumn + '` asc);';

			let mysqlCmd = mysql.format(sql);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, result) => {

					if (err) {
						rej(err);
						return;
					}

					res();
				})
			});
		}

		dao.migrateDbmsUsersAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate DBMS Users', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			let sql = 'create user \'' + dao._dbConfig.conn.user + '\'';
			sql += ' identified by \'' + dao._dbConfig.conn.password + '\' password expire never;';

			sql += ' grant create temporary tables, create view, delete, execute, insert, select, update';
			sql += ' on `' + dao._dbConfig.conn.database + '`.* to \'' + dao._dbConfig.conn.user + '\'';

			let mysqlCmd = mysql.format(sql);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, result) => {

					if (err) {
						rej(err);
						return;
					}

					res();
				});
			});
		};

		dao.migrateTestResultAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Test Result', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			let sql = 'create table `' + schema.testResult.table + '`';
			sql += ' (`' + schema.testResult.testIdColumn + '` varchar(10) not null,';
			sql += ' `' + schema.testResult.languageColumn + '` varchar(50) null,';
			sql += ' `' + schema.testResult.dataColumn + '` json not null,';

			sql += ' constraint ' + schema.testResult.testResultTestIndex;
			sql += ' unique(`' + schema.testResult.testIdColumn + '`,`' + schema.testResult.languageColumn + '`));';

			let mysqlCmd = mysql.format(sql);

			return new Promise((res, rej) => {

				dao._client.query(mysqlCmd, (err, result) => {

					if (err) {
						rej(err);
						return;
					}

					res();
				})
			});
		}

		dao.getLatestVersion = () => { return '1.0.0'; };

		return new Promise((res, rej) => {

			dao._openConnectionAsync()
				.then(() => {

					dao._createDatabaseAsync()
						.then(() => {

							dao._createSchemaVersionAsync()
								.then(() => {

									res(dao);
								})
								.catch(rej);
						})
						.catch(rej);
				})
				.catch(rej);
		})
	}
}